upgrades/steps_28.go

```diff
@@ -321,6 +325,137 @@ func readStorageStateFile(dirPath string, tag names.StorageTag) (bool, error) {
 	return *info.Attached, nil
 }
 
+// storageDiskInfo defines the storage unit storageData serialization.
 type storageDiskInfo struct {
 	Attached *bool `yaml:"attached,omitempty"`
 }
+
+type relationData struct {
+	data     map[int]string
+	filename string
+}
+
+func readRelationState(uniterStateDir string) (relationData, error) {
+	relationStateDir := filepath.Join(uniterStateDir, "relations")
+
+	rSt, err := readAllRelationStateDirs(relationStateDir)
+	if err != nil {
+		return relationData{}, err
+	}
+	retData := relationData{data: make(map[int]string, len(rSt)), filename: relationStateDir}
+	for k, v := range rSt {
+		dataYaml, err := yaml.Marshal(v)
+		if err != nil {
+			return relationData{}, errors.Annotatef(err, "failed to unmarshal relation state from %q", relationStateDir)
+		}
+		retData.data[k] = string(dataYaml)
+	}
+	return retData, nil
+}
+
+// readAllRelationStateDirs loads and returns every relationState persisted directly inside
+// the supplied dirPath. If dirPath does not exist, no error is returned.
+func readAllRelationStateDirs(dirPath string) (dirs map[int]*relationState, err error) {
+	defer errors.DeferredAnnotatef(&err, "cannot load relations state from %q", dirPath)
+	if _, err := os.Stat(dirPath); os.IsNotExist(err) {
+		return nil, nil
```

Comment:
If the dir does not exist, we return nil, and `readRelationState` returns back an empty map. Is this the intended behavior?